### PR TITLE
Standalone Executable

### PR DESCRIPTION
**Pull request by artoonie**
Auto-create executables using jpackage.

Issues:
1. The version number is now in three different places. We need an easy way to sync this across the project, but I couldn't find a straightforward way yet. We'll need a ticket to track this.
2. We're now creating a ton of packages: the launchable binaries (jpackage), the java runtime image (jlinkZip), SHAs for each, x3 platforms --> the packages are around 350mb. When a release is created, this won't be an issue since people can just download the relevant file, but we should consider whether we want to support the jlinkZip image indefinitely, or if the jpackage will supercede it

### CODE REVIEW

**Comment by HEdingfield** [comment]
Curious why we're merging this into `master` instead of `develop`? Just so we can test it out sooner?

> MacOS packages aren't signed, and seem impossible to install on newer macs. They are painful to install on older macs. We probably need to [sign the packages](https://docs.github.com/en/actions/deployment/deploying-xcode-applications/installing-an-apple-certificate-on-macos-runners-for-xcode-development), which I believe require paying Apple $100/year?

My knee-jerk reaction to this is helllllllll no. But I guess we can see what @chughes297 thinks.

Also, I'm happy to test the bins for Linux and Windows if you can point me to where they are.

> The version number is now in three different places. We need an easy way to sync this across the project, but I couldn't find a straightforward way yet. We'll need a ticket to track this.

We should be able to get rid of it in `release.yml`, right, defaulting to the GitHub tag variable instead? Looks like there are some merge conflicts to resolve with `release.yml` anyway.

> We're now creating a ton of packages: the launchable binaries (jpackage), the java runtime image (jlinkZip), SHAs for each, x3 platforms --> the packages are around 350mb. When a release is created, this won't be an issue since people can just download the relevant file, but we should consider whether we want to support the jlinkZip image indefinitely, or if the jpackage will supercede it

Yeah, I think jpackage will replace jlinkZip. However, we should discuss the ramifications of this with @chughes297 to make sure it doesn't cause any additional security issues having them use a compiled version. Also wondering if we need to provide SHA512 checksums for the bins too.

**Comment by chughes297** [comment]
>> MacOS packages aren't signed, and seem impossible to install on newer macs. They are painful to install on older macs. We probably need to [sign the packages](https://docs.github.com/en/actions/deployment/deploying-xcode-applications/installing-an-apple-certificate-on-macos-runners-for-xcode-development), which I believe require paying Apple $100/year?
> 
> My knee-jerk reaction to this is helllllllll no. But I guess we can see what @chughes297 thinks.
> 
Why hell no? I need to confirm this but I assume RCVRC can cover those costs, especially if it makes install that much easier. 
> Also, I'm happy to test the bins for Linux and Windows if you can point me to where they are.
> 
>> We're now creating a ton of packages: the launchable binaries (jpackage), the java runtime image (jlinkZip), SHAs for each, x3 platforms --> the packages are around 350mb. When a release is created, this won't be an issue since people can just download the relevant file, but we should consider whether we want to support the jlinkZip image indefinitely, or if the jpackage will supercede it
> 
> Yeah, I think jpackage will replace jlinkZip. However, we should discuss the ramifications of this with @chughes297 to make sure it doesn't cause any additional security issues having them use a compiled version. Also wondering if we need to provide SHA512 checksums for the bins too.

I need a bit more info about what is being discussed - not sure I understand what the potential security issues are if using a compiled version? And yes I believe we'll need to continue providing checksums for each release (if that's what's being asked lol).

File: src/main/java/network/brightspots/rcv/Main.java
Comment (HEdingfield) [inline src/main/java/network/brightspots/rcv/Main.java:29]: Any ideas on how we can create a common spot to access these vars for both `Main.java` and `build.gradle`?

I don't think it's a deal-breaker if we can't, as I typically just Ctrl-Shift-F in IntelliJ and search for the current version number any time I want to update it (which would catch both spots). Probably safe to get rid of both TODOs in that case, and maybe create an issue for the backlog if you think it's worth it.

File: build.gradle
Comment (HEdingfield) [inline build.gradle:101]: Nit: I lean towards keeping this as-is since it matches the option `add-modules`, and it also hints that we can add more if necessary.

File: build.gradle
Comment (HEdingfield) [inline build.gradle:122]: Nit: I think we can get rid of "The" in here.

File: .github/workflows/release.yml
Comment (HEdingfield) [inline .github/workflows/release.yml:1]: Looks like merge conflicts in here -- I'll look it over once those are resolved.

File: .github/actions/gradle-and-sha/action.yml
Comment (HEdingfield) [inline .github/actions/gradle-and-sha/action.yml:1]: Nice refactor!

File: .github/actions/gradle-and-sha/action.yml
Comment (HEdingfield) [inline .github/actions/gradle-and-sha/action.yml:21]: Nit: Rename output file

File: .github/actions/gradle-and-sha/action.yml
Comment (HEdingfield) [inline .github/actions/gradle-and-sha/action.yml:23]: Is `jpackage` supposed to be explicitly referenced in this refactored action?

**Comment by HEdingfield** [comment]
> @chughes297: Why hell no? I need to confirm this but I assume RCVRC can cover those costs, especially if it makes install that much easier.

I guess because I'm allergic to costs associated with open source software (especially recurring ones) but, more importantly, I thought all the election tabulation stations ran Windows anyway?

> @chughes297: I need a bit more info about what is being discussed - not sure I understand what the potential security issues are if using a compiled version? And yes I believe we'll need to continue providing checksums for each release (if that's what's being asked lol).

I guess I was thinking because it's just an extra level of opacity, which probably also creates a registry entry on their machine (correct me if I'm wrong, @artoonie).

File: src/main/java/network/brightspots/rcv/Main.java
Comment (artoonie) [reply src/main/java/network/brightspots/rcv/Main.java:29]: I think having a standalone `VERSION` or `version.txt` would be sufficient, but I need to read more about best practices in Java/Gradle to make that call.

My plan was to (1) create a ticket, (2) add the ticket number to the TODO. Whenever I need to copy/paste changes across a codebase, I like to also copy/paste a comment nearby that tells me all the places it needs to be synced.

**Comment by chughes297** [comment]
> > @chughes297: Why hell no? I need to confirm this but I assume RCVRC can cover those costs, especially if it makes install that much easier.
> 
> I guess because I'm allergic to costs associated with open source software (especially recurring ones) but, more importantly, I thought all the election tabulation stations ran Windows anyway?
> 
Most do. So far in my experience, though, some offices have a Linux machine handy and others may have a Mac workstation they can use. I lean towards being as comprehensive as possible with OSes so we can be flexible and jurisdictions can ideally use a piece of hardware they already have sitting around instead of having to buy a new one. 
>>I guess I was thinking because it's just an extra level of opacity, which probably also creates a registry entry on their machine (correct me if I'm wrong, @artoonie).
>
>Yes, that is true -- when installing the software, you cede some control to the operating system which could be seen as reducing transparency. For example: If you try installing 1.3.0 but you previously had 1.4.0 installed, will your operating system decline to downgrade, but show you that it succeeded? The answer is No today, but that's the type of thing we lose control over.
>
>(This may not be the best example of what can go wrong, but it's the only thing I can think of right now.)
>
>However, assuming other election software follows the same model (which I expect they do), I think we can feel safe following the same best practices?
I'm curious to hear more about the security/opacity argument (maybe easier to talk out on the phone?). And I would need to ask vendors what their standard practice is but I imagine it is similar to what Armin described.

File: .github/workflows/release.yml
Comment (HEdingfield) [reply .github/workflows/release.yml:1]: Merge conflicts still need resolving, even after branch change.

File: src/main/java/network/brightspots/rcv/Main.java
Comment (HEdingfield) [reply src/main/java/network/brightspots/rcv/Main.java:29]: That sounds perfect.

**Review by HEdingfield — COMMENTED** [review]
>> Also, I'm happy to test the bins for Linux and Windows if you can point me to where they are.
>
> https://github.com/BrightSpots/rcv/actions/runs/4801866529

Unfortunately I can't install this package on my Linux machine due to org policy. We'll have to find another way to test (e.g. VirtualBox).

I was, however, able to verify that the zip built by the current release.yml works on Linux though!

File: .github/workflows/release.yml
Comment (artoonie) [reply .github/workflows/release.yml:1]: Yeah, the merge conflict is due to `develop` removing build-on-branches and this branch adding it back so I can test it on every push. I didn't plan to resolve until all testing was complete, since it's a nice reminder to delete that line and prevents us from accidentally keeping it. I can resolve if you prefer though, there are other ways to make sure it doesn't get merged (e.g. a Request Changes review on GH).

File: .github/actions/gradle-and-sha/action.yml
Comment (artoonie) [reply .github/actions/gradle-and-sha/action.yml:23]: No, this was just debugging code since every OS spat out several file options, so this was a way to inspect and see which we wanted (e.g. Windows has MSI and EXE; Mac has DMG and .App). I've removed it now.

File: build.gradle
Comment (artoonie) [reply build.gradle:101]: Agreed. For some reason, I thought the modules would be space-separated rather than comma-separated, so I thought only one module was allowed. I'll update the comment for future travelers.

File: src/main/java/network/brightspots/rcv/Main.java
Comment (artoonie) [reply src/main/java/network/brightspots/rcv/Main.java]: https://github.com/BrightSpots/rcv/issues/662 and also noted in the comments now, feel free to add any context needed to that issue

File: .github/workflows/release.yml
Comment (HEdingfield) [reply .github/workflows/release.yml:1]: Ok, I'll just review by diffing in IntelliJ against this file, but I'll hold off on doing the final approval and merge until the conflict is resolved, because I think it's a good idea to have another pair of eyes when resolving merge conflicts.

File: .github/workflows/release.yml
Comment (HEdingfield) [inline .github/workflows/release.yml:37]: In my failed Linux check today, the lack of extension was a little confusing, but the OS seemed to at least recognize it was a package when I tried to run it (and rejected it due to security policy, as mentioned). Hopefully it's correct to keep it without an extension (e.g. no `.deb`). Will leave it up to your Virtualbox testing to verify.

File: .github/workflows/release.yml
Comment (HEdingfield) [inline .github/workflows/release.yml:45]: Can you leave a comment about why these numbers need to be hard-coded in here? I have a feeling the question will come up repeatedly in the future.

File: .github/workflows/release.yml
Comment (HEdingfield) [inline .github/workflows/release.yml:100]: Is it only uploading the jlinkzip version since this is zipfn here? I think we'd also want to upload the jpackage versions.

I also noticed that the artifact file you linked to for Linux testing was a single file with no extension that turned out to be a zip with every single platform and SHA in it. Hopefully we're not uploading something like that to the release, as it was fairly confusing. I'd prefer individual files as the current release.yml does.

File: src/main/java/network/brightspots/rcv/Main.java
Comment (HEdingfield) [inline src/main/java/network/brightspots/rcv/Main.java:29]: Here and in the other two spots: I recommend linking the full URL just in case this gets forked, e.g.:

// TODO: Sync these values with release.yml and build.gradle (https://github.com/BrightSpots/rcv/issues/625)

**Comment by artoonie** [comment]
@chughes297 the logo needs to be a perfect square. The two logos I see are either slightly landscape or slightly portrait. Of these three variations, which do you prefer? Or something in-between? Whitespace would not look great on the logo, thus Option 3, but Option 3 doesn't match your other iconography.

image
image
image

File: .github/workflows/release.yml
Comment (artoonie) [reply .github/workflows/release.yml:100]: 1. Before we merge this, we need to decide whether we want to replace jlinkzip with jpackage, release both, or continue just releasing jlinkzip until we've had more time/feedback on jpackage. I believe that decision is [pending feedback from vendors](https://github.com/BrightSpots/rcv/pull/660#issuecomment-1523536314).
2. Good catch on the linux release! No, that's not intentional. I'll fix it.

File: .github/workflows/release.yml
Comment (artoonie) [reply .github/workflows/release.yml]: I've corrected this: previously, I just shipped the binary (thus no extension), but changed it to use .deb now and forgot to sync that change.

The .deb seems to be correct -- however, I haven't booted my linux partition in several years and it seems like testing on my end will be pretty hard. The machine has no access to the internet (seems like the wifi hardware is too new), and when manually copying the .deb via a USB stick, it's unable to open it because it uses a newer compression format not supported by Ubuntu 16.04.

All this to say: if somebody else has faster access to Linux, could you test it fully? Otherwise, I'll need to install virtualbox or reformat my linux partition, both of which will take some time to set up.

File: .github/workflows/release.yml
Comment (HEdingfield) [reply .github/workflows/release.yml:101]: Oh, and I didn't notice the RCVRC logo at any point. The installer just had a red CD-looking thing, and the program icon was the Java Star Trek looking icon.

File: .github/workflows/release.yml
Comment (artoonie) [reply .github/workflows/release.yml]: Ah, that does remind me-- I've got some Raspberry Pis ready-to-go laying around. I'll see if I can boot one of those up.

File: .github/workflows/release.yml
Comment (artoonie) [reply .github/workflows/release.yml:101]: Ah, I understand: yes, that is a limitation of Github Artifacts on Actions. In a release, they will be individual files, but Github forces them all into one giant package on Github Actions Artifacts.

i.e. this won't be an issue when we create releases, only in development.

**Comment by HEdingfield** [comment]
I believe this would resolve #57 as well.

**Comment by artoonie** [comment]
> I believe this would resolve #57 as well.

Indeed :) that's the branch name.

**Comment by HEdingfield** [comment]
> > I believe this would resolve #57 as well.
> 
> Indeed :) that's the branch name.

Ah yeah! I was just tagging the issue in GH so this PR would actually appear in its history to help keep things organized.

File: .github/workflows/release.yml
Comment (artoonie) [reply .github/workflows/release.yml]: i've reinstalled a more recent Ubuntu and confirmed that this works, though the install doesn't put the binary in the $PATH -- you have to manually launch it from `/opt/`. I can't figure out how to make the package more easily launchable, so I'm happy to call this "good enough" for now given the lack of linux users, and the easy workaround. I've added info to the README. We can revisit later if we want to make this simpler, but for now, just run `/opt/rctab/bin/RCTab` to launch after installing the .deb.

**Comment by artoonie** [comment]
This is almost ready for review. I've tested on Linux and added an app icon since the last convo. Last step will be to sign the mac application.

I am still worried about this never getting tested until we run a release -- but, as requested, I have set this to only run on release.

**Comment by HEdingfield** [comment]
Nice! Shout out when you think it's ready.

> I am still worried about this never getting tested until we run a release -- but, as requested, I have set this to only run on release.

I have no issue at all with temporarily making it run during events other than release for testing purposes. Then we can just revert it to run only on release once we're satisfied.

File: build.gradle
Comment (HEdingfield) [inline build.gradle:118]: I assume it was 1.3.1 just for testing purposes? Can we set to 1.4.0 now?

File: build.gradle
Comment (HEdingfield) [inline build.gradle:169]: I appreciate how painful it must have been to get all these params just right!

File: .github/workflows/release.yml
Comment (HEdingfield) [inline .github/workflows/release.yml:129]: I think I asked about this previously but forgot the answer... I see in the test run [here](https://github.com/BrightSpots/rcv/actions/runs/5213169580) that it creates a single file called "Package", with all three versions in it. Is it possible to upload the individual versions and their hashes separately instead of all bundled together?

File: .github/workflows/release.yml
Comment (HEdingfield) [inline .github/workflows/release.yml:8]: Do we update this to `master`?

File: .github/workflows/release.yml
Comment (HEdingfield) [inline .github/workflows/release.yml:49]: Update 1.3.1 in these three spots to 1.4.0 as well?

File: .github/workflows/release.yml
Comment (HEdingfield) [inline .github/workflows/release.yml:48]: I noticed the file I downloaded inside `Package.zip` is called `rctab_feature_issue-57_standalone-executable_Windows.exe`, which seems off... it should be the version number, shouldn't it?

Comment (HEdingfield) [inline .github/workflows/release.yml:48]: Not a huge deal, but the start menu icon in Windows still has the Java icon instead of the launcher icon you're including in this PR. Not sure if that's possible to fix easily?

**Review by HEdingfield — COMMENTED** [review]
Some light changes for you [here](https://github.com/BrightSpots/rcv/pull/660/commits/af18a07eba4d9b0df0ace77a9cf115beb6ffbec0) to review too, per our discussion.

File: src/main/java/network/brightspots/rcv/Main.java
Comment (artoonie) [reply src/main/java/network/brightspots/rcv/Main.java:33]: `jpackage` required in one of the OSs, I forget which, maybe all. It's standard semantic versioning to use a dash instead of a dot here: https://semver.org/

File: build.gradle
Comment (artoonie) [reply build.gradle:118]: I'm a little afraid to set it to 1.4.0 prematurely, as we'd have fake but signed `1.4.0` versions floating around. I guess the same is true with `1.3.1`?

The issue stems from the fact that Mac won't allow a bundle to have a `-alpha` at the end, so we need some way of signifying a fake/development version number. See: https://stackoverflow.com/questions/70066384/javafx-gradle-jlink-plugin-jpackageimage-fails-because-of-1-0-snapshot-conta

Perhaps we should use 0.0.0 unless we are building a release? Or maybe this just isn't a real issue, and we're okay having "false" 1.4.0s around?

File: build.gradle
Comment (artoonie) [reply build.gradle:169]: The a-ha moment was realizing I need two different keys, one for the installer, one for the application. *headdesk*

File: .github/workflows/release.yml
Comment (artoonie) [reply .github/workflows/release.yml:129]: Only in a Release, but not in an Artifact. So here, it would be possible: https://github.com/BrightSpots/rcv/releases/tag/

And it is indeed what happens there -- just not in the Artifact.

File: .github/workflows/release.yml
Comment (artoonie) [reply .github/workflows/release.yml:8]: I think we delete this entirely when we're ready to merge, and create a Release instead.

File: .github/workflows/release.yml
Comment (artoonie) [reply .github/workflows/release.yml:48]: It will be in the final release! Just not in development builds, since it uses the ref name. The ref name is the branch name for these dev builds, and the release name for release builds.

Comment (artoonie) [reply .github/workflows/release.yml:48]: Done! Windows needed a `ico` file instead of `png`.

File: .github/workflows/release.yml
Comment (HEdingfield) [reply .github/workflows/release.yml:49]: Let's update all these to 1.4.0-alpha too.

File: .github/workflows/release.yml
Comment (HEdingfield) [reply .github/workflows/release.yml:8]: SGTM, let's keep this convo open and I'll make my final approval contingent upon deleting this.

File: build.gradle
Comment (HEdingfield) [reply build.gradle:118]: Dumb Apple. That may be a blessing in disguise in a small way, though, because we don't want to be creating these builds / releasing alpha versions. I'd say we set it to `1.4.0-alpha` here now so the find-replace is easier later (plus I guess it'll block building).

File: src/main/java/network/brightspots/rcv/Main.java
Comment (HEdingfield) [reply src/main/java/network/brightspots/rcv/Main.java:33]: Fair point. I think we're going to need to update `isVersionNewer` in this PR too though if you want to do this here now. Alternatively, you could leave it as-is and we can file another P1 issue to deal with this.

File: .github/workflows/release.yml
Comment (HEdingfield) [reply .github/workflows/release.yml:48]: Good catch. I'll test this out and verify one last time before final approval.

File: src/main/java/network/brightspots/rcv/Main.java
Comment (artoonie) [reply src/main/java/network/brightspots/rcv/Main.java:33]: https://github.com/BrightSpots/rcv/issues/700 i'll deal with it separately

File: .github/workflows/release.yml
Comment (HEdingfield) [reply .github/workflows/release.yml:49]: Worked perfectly!

**Review by HEdingfield — APPROVED** [review]
Fantastic job! This would have taken me forever to get going.

**Pull request merged**
